File: InsuraTech.Domain/Common/ValueObject.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace InsuraTech.Domain.Common
{
    public abstract class ValueObject
    {
        protected abstract IEnumerable<object> GetEqualityComponets();

        public override bool Equals(object? obj)
        {
            if (obj is null || obj.GetType() != GetType()) return false;
            return GetEqualityComponets()
                    .SequenceEqual(((ValueObject)obj).GetEqualityComponets());
            
        }
        public override int GetHashCode() =>
            GetEqualityComponets()
            .Aggregate(1, (current, obj) =>
                HashCode.Combine(current, obj?.GetHashCode() ?? 0));

        public static bool operator ==(ValueObject? left, ValueObject? right) =>
                left?.Equals(right) ?? right is null;

        public static bool operator !=(ValueObject? left, ValueObject? right) =>
            !(left == right);
    }
}

File: backend_v2/src/main/java/com/insuratech/domain/common/ValueObject.java
package com.insuratech.domain.common;

public abstract class ValueObject {

    protected abstract Object[] getEqualityComponents();

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        ValueObject other = (ValueObject) obj;
        Object[] components = getEqualityComponents();
        Object[] otherComponents = other.getEqualityComponents();
        if (components.length != otherComponents.length) return false;
        for (int i = 0; i < components.length; i++) {
            if (components[i] == null && otherComponents[i] == null) continue;
            if (components[i] == null || !components[i].equals(otherComponents[i])) return false;
        }
        return true;
    }

    @Override
    public int hashCode() {
        int result = 1;
        for (Object component : getEqualityComponents()) {
            result = 31 * result + (component == null ? 0 : component.hashCode());
        }
        return result;
    }
}
